Src/VimMac/InlineRename/InlineRenameListenerFactory.cs

```diff
@@ -0,0 +1,127 @@
+            IsActive;
+
+        internal IInlineRenameUtil RenameUtil
+        {
+            get { return _inlineRenameUtil; }
+            set
+            {
+                if (_inlineRenameUtil != null)
+                {
+                    _inlineRenameUtil.IsRenameActiveChanged -= OnIsRenameActiveChanged;
+                }
+
+                _inlineRenameUtil = value;
+                if (_inlineRenameUtil != null)
+                {
+                    _inlineRenameUtil.IsRenameActiveChanged += OnIsRenameActiveChanged;
+                }
+            }
+        }
+
+        /// <summary>
+        /// The inline rename utility manipulates the undo / redo buffer
+        /// directly during a  rename.  Need to register this as expected so
+        /// the undo implementation doesn't raise any errors.
+        /// </summary>
+        internal bool IsActive => _inlineRenameUtil != null && _inlineRenameUtil.IsRenameActive;
+
+        internal bool InRename { get; private set; }
+
+        [ImportingConstructor]
+        internal InlineRenameListenerFactory(
+            IVimApplicationSettings vimApplicationSettings)
+        {
+            _vimApplicationSettings = vimApplicationSettings;
+
+            if (InlineRenameUtil.TryCreate(out IInlineRenameUtil renameUtil))
+            {
+                RenameUtil = renameUtil;
+            }
+        }
+
+        private void OnIsRenameActiveChanged(object sender, EventArgs e)
+        {
+            if (InRename && !_inlineRenameUtil.IsRenameActive)
+            {
+                InRename = false;
+                foreach (var vimBuffer in _vimBufferList)
+                {
+                    vimBuffer.SwitchedMode -= OnModeChange;
+                    if (vimBuffer.ModeKind == ModeKind.ExternalEdit)
+                    {
+                        vimBuffer.SwitchPreviousMode();
+                    }
+                }
+            }
+            else if (!InRename && _inlineRenameUtil.IsRenameActive)
+            {
+                InRename = true;
+                foreach (var vimBuffer in _vimBufferList)
+                {
+                    // Respect the user's edit monitoring setting.
+                    if (_vimApplicationSettings.EnableExternalEditMonitoring)
+                    {
+                        vimBuffer.SwitchMode(ModeKind.ExternalEdit, ModeArgument.None);
+                    }
+                    vimBuffer.SwitchedMode += OnModeChange;
+                }
+            }
+        }
+
+        private void OnModeChange(object sender, SwitchModeEventArgs args)
+        {
+            if (InRename && _inlineRenameUtil.IsRenameActive && args.ModeArgument.IsCancelOperation)
+            {
+                _inlineRenameUtil.Cancel();
+            }
+        }
+
+        internal static bool IsInlineRenameContentType(IContentType contentType)
+        {
+            return
+                contentType.IsCSharp() ||
+                //contentType.IsFSharp() ||

```

Comment:
Just wanted to make sure this was deliberate as VSMac supports F#. It's not a blocker either way.

Reply:
Can't remember why I did this. I think it was because VSMac has no support for new editor and F#. But I am working on that :)

I think I should uncomment this.